### PR TITLE
Fixed #32256 -- Fixed migration optimization crash when swapping field names.

### PR DESCRIPTION
**Pull request by hramezani**
https://code.djangoproject.com/ticket/32256

### CODE REVIEW

**Review by charettes — COMMENTED** [review]
This should be testable at the optimizer level without involving on disk migration files.

See `test_optimizer` instead of `test_command`

**Comment by hramezani** [comment]
@charettes Thanks for your suggestion. I moved the test to `test_optimizer`.

**Review by InvalidInterrupt — COMMENTED** [review]
What if the forwards/backwards functions in a `RunPython` operation referenced the renamed fields (by their old name, because it was written before the rename operations)?
Also, I believe it likely the same crash could also be caused using other operations that present an optimisation barrier (e.g. `SeparateDatabaseAndState` or `RunSQL`) in place of `RunPython`.

**Comment by hramezani** [comment]
> What if the forwards/backwards functions in a RunPython operation referenced the renamed fields (by their old name, because it was written before the rename operations)?

I am not sure how we can handle this case.

> Also, I the same crash could also be caused using other operations that present an optimisation barrier (e.g. SeparateDatabaseAndState (tested and confirmed) or RunSQL) in place of RunPython.

We can add the `reduce` function for these operations as well.

@felixxm @charettes What do you think?

**Comment by InvalidInterrupt** [comment]
Thanks to having your test as reference I put together my idea for fixing this in `RenameField.reduce()` here:
https://github.com/InvalidInterrupt/django/tree/ticket_32256.

I think it's better to fix this there rather than try to prevent the optimizer from considering such optimizations in the first case; other proposed features I've seen in trac (i.e. https://code.djangoproject.com/ticket/27844) would be likely to cause the issue to arise again. My approach does assume that this sequence of migrations is not optimizable; I would be happy to be proven wrong if someone identified a way to optimize this in the general case.

I'm (probably overly) concerned about stepping on toes here, so I'm not making another PR. I may find myself busy again over the next couple days; please feel free to either ask me to make a PR or just take my commit if that's easier for you all.

**Review by AlexisMunera98 — APPROVED** [review]

**Comment by felixxm** [comment]
@InvalidInterrupt proposition to avoid optimization sounds reasonable, but I'm not sure if the fix is in the right place :thinking:

Simon, can I ask for your opinion? (\cc @charettes)

File: tests/migrations/test_optimizer.py [hunk 1]
```diff
@@ -830,3 +830,30 @@ def test_optimize_elidable_operation(self):
...
+            [
+                migrations.CreateModel(
+                    'MyModel',
+                    [
+                        ('field_b', models.IntegerField()),
+                        ('field_a', models.IntegerField()),
+                    ],
+                ),
+                migrations.RunPython(migrations.RunPython.noop),
+            ],
```
Comment (charettes) [inline tests/migrations/test_optimizer.py:858]: This optimization can't be allowed.

What if `RunPython` pointed at a function that referenced specific fields? e.g.


```python
def forwards(apps, schema_editor):
    apps.get_model('app.MyModel').objects.filter(field_a__lte=42).update(field_a=14)

operations = [
    migrations.CreateModel(
        'MyModel',
        [
            ('field_a', models.IntegerField()),
            ('field_b', models.IntegerField()),
        ],
    ),
    migrations.RunPython(forwards),
    migrations.RenameField('MyModel', 'field_a', 'field_c'),
]
```

`RunPython` opaque operations that must be optimization barriers as they can refer to model definitions at any time. They require either to be marked as `elidable` if they can be removed or adjusted by users on migration squashing.

**Review by charettes — COMMENTED** [review]
I didn't look into details at https://github.com/InvalidInterrupt/django/commit/24e1e7386b9a46d4ca92686b03b7600667d2c77b from @InvalidInterrupt but it's definitely a more appropriate approach than this one. I'd be interested to see the suite running against it.

To me this is the same problem as trying to swap two variables values assuming your _language_ doesn't have support for such an operation

```python
a = 1
b = 2
c = a
a = b
b = c
```

There's no way to optimize through this sequence of operations unless we add a `RenameFields` operation so it definitely looks like an issue with `RenameField.reduce`.

**Comment by hramezani** [comment]
@charettes @felixxm I've updated the PR with @InvalidInterrupt patch.

**Review by charettes — APPROVED** [review]
Thanks @hramezani, that looks good to me.

File: tests/migrations/test_optimizer.py [hunk 2]
```diff
@@ -866,3 +866,33 @@ def test_optimize_elidable_operation(self):
                 migrations.CreateModel("Phou", [("name", models.CharField(max_length=255))]),
             ],
         )
+
+    def test_optimizer_crashes_when_fields_names_are_swapped(self):
+        self.assertOptimizesTo(
```
Comment (felixxm) [inline tests/migrations/test_optimizer.py:871]:
```suggestion
        self.assertDoesNotOptimize(
```

**Review by felixxm — COMMENTED** [review]
@hramezani Thanks :+1: I pushed minor edits.

**Pull request merged**